src/main/java/org/opensearch/flowframework/template/ProcessNode.java

```diff
@@ -0,0 +1,184 @@
+     *
+     * @param id A string identifying the workflow step
+     * @param workflowStep A java class implementing {@link WorkflowStep} to be executed when it's this node's turn.
+     */
+    ProcessNode(String id, WorkflowStep workflowStep) {
+        this(id, workflowStep, WorkflowData.EMPTY);
+    }
+
+    /**
+     * Create this node linked to its executing process.
+     *
+     * @param id A string identifying the workflow step
+     * @param workflowStep A java class implementing {@link WorkflowStep} to be executed when it's this node's turn.
+     * @param input Input required by the node
+     */
+    public ProcessNode(String id, WorkflowStep workflowStep, WorkflowData input) {
+        this.id = id;
+        this.workflowStep = workflowStep;
+        this.input = input;
+    }
+
+    /**
+     * Returns the node's id.
+     * @return the node id.
+     */
+    public String id() {
+        return id;
+    }
+
+    /**
+     * Returns the node's workflow implementation.
+     * @return the workflow step
+     */
+    public WorkflowStep workflowStep() {
+        return workflowStep;
+    }
+
+    /**
+     * Returns the input data for this node.
+     * @return the input data
+     */
+    public WorkflowData input() {
+        return input;
+    }
+
+    /**
+     * Returns a {@link CompletableFuture} if this process is executing.
+     * Relies on the node having been sorted and executed in an order such that all predecessor nodes have begun execution first (and thus populated this value).
+     *
+     * @return A future indicating the processing state of this node.
+     * Returns {@code null} if it has not begun executing, should not happen if a workflow is sorted and executed topologically.
+     */
+    public CompletableFuture<WorkflowData> getFuture() {
+        return future;
+    }
+
+    /**
+     * Returns the predecessors of this node in the workflow.
+     * The predecessor's {@link #getFuture()} must complete before execution begins on this node.
+     *
+     * @return a set of predecessor nodes, if any.  At least one node in the graph must have no predecessors and serve as a start node.
+     */
+    public Set<ProcessNode> getPredecessors() {
+        return predecessors;
+    }
+
+    /**
+     * Sets the predecessor node.  Called by {@link TemplateParser}.
+     *
+     * @param predecessors The predecessors of this node.
+     */
+    void setPredecessors(Set<ProcessNode> predecessors) {
+        this.predecessors = Set.copyOf(predecessors);
+    }
+
+    /**
+     * Execute this node in the sequence. Initializes the node's {@link CompletableFuture} and completes it when the process completes.
+     *
+     * @return this node's future. This is returned immediately, while process execution continues asynchronously.
+     */
+    public CompletableFuture<WorkflowData> execute() {
+        this.future = new CompletableFuture<>();
+        CompletableFuture.runAsync(() -> {
```

Comment:
This is where the thread is leaked. 
>Returns a new CompletableFuture that is asynchronously completed by a task running in the ForkJoinPool.commonPool() after it runs the given action.

I would suggest executing it bounded using a defined executor. You can use - 
```
runAsync(Runnable runnable, Executor executor)
```
The executor can be either passed in or defined as a class variable.

The tests can then do something similar to https://github.com/opensearch-project/OpenSearch/blob/d34b35272fa19bc53cb0f6843a92d2d1ee4e3b47/server/src/test/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListenerTests.java#L50-L58

Reply:
The ProcessNode class currently doesn't take a thread pool as an argument on instantiation.   Set the thread leak scope to NONE as a temporary fix.  Created issue https://github.com/opensearch-project/opensearch-ai-flow-framework/issues/42 to track permanent fix which will allow implementation of this approach for tests and properly using OpenSearch thread pool in production.